src/ballet/pack/fd_pack.c

```diff
@@ -0,0 +1,620 @@
+  void * _txnq       = FD_SCRATCH_ALLOC_APPEND( l, txnq_align(),                   txnq_footprint( pack_depth )           );
+  void * _freelist   = FD_SCRATCH_ALLOC_APPEND( l, freelist_align(),               freelist_footprint( freelist_cnt )     );
+  void * _uses_read  = FD_SCRATCH_ALLOC_APPEND( l, acct_uses_align(),              acct_uses_footprint( lg_uses_tbl_sz )  );
+  void * _uses_write = FD_SCRATCH_ALLOC_APPEND( l, acct_uses_align(),              acct_uses_footprint( lg_uses_tbl_sz )  );
+
+  pack->bank_cnt   = bank_cnt;
+  pack->bank_depth = bank_depth;
+  pack->pack_depth = pack_depth;
+  pack->est_tbl    = est_tbl;
+  pack->rng        = rng;
+
+  fd_memset( _bank_status, 0, sizeof(fd_pack_bank_status_t)*bank_cnt );
+
+  outq_new(      _outq,       bank_cnt       );
+  txnq_new(      _txnq,       pack_depth     );
+  freelist_new(  _freelist,   freelist_cnt   );
+  acct_uses_new( _uses_read,  lg_uses_tbl_sz );
+  acct_uses_new( _uses_write, lg_uses_tbl_sz );
+
+
+  /* Populate the freelist */
+  fd_txn_p_t * * freelist = freelist_join( _freelist );
+  ulong ptr = (ulong)txnmem;
+  for( ulong i=0UL; i<freelist_cnt; i++ ) {
+    ptr = fd_ulong_align_up( ptr, fd_pack_txnmem_align() );
+    freelist_push_tail( freelist, (fd_txn_p_t *)ptr );
+    ptr += sizeof(fd_txn_p_t);
+  }
+  freelist_leave( freelist );
+
+  return mem;
+}
+
+fd_pack_t *
+fd_pack_join( void * mem ) {
+  FD_SCRATCH_ALLOC_INIT( l, mem );
+  fd_pack_t * pack  = FD_SCRATCH_ALLOC_APPEND( l, FD_PACK_ALIGN, sizeof(fd_pack_t) );
+  ulong bank_cnt   = pack->bank_cnt;
+  ulong bank_depth = pack->bank_depth;
+  ulong pack_depth = pack->pack_depth;
+
+  ulong freelist_cnt   = bank_depth*bank_cnt + pack_depth + bank_cnt;
+  int   lg_uses_tbl_sz = fd_ulong_find_msb( fd_ulong_pow2_up( 2UL*FD_TXN_ACCT_ADDR_MAX*bank_cnt ) );
+
+  const ulong bank_status_align = alignof(fd_pack_bank_status_t);
+  pack->bank_status     =                 FD_SCRATCH_ALLOC_APPEND( l, bank_status_align, sizeof(fd_pack_bank_status_t)*bank_cnt );
+  pack->outq            = outq_join(      FD_SCRATCH_ALLOC_APPEND( l, outq_align(),      outq_footprint( bank_cnt )           ) );
+  pack->txnq            = txnq_join(      FD_SCRATCH_ALLOC_APPEND( l, txnq_align(),      txnq_footprint( pack_depth )         ) );
+  pack->freelist        = freelist_join(  FD_SCRATCH_ALLOC_APPEND( l, freelist_align(),  freelist_footprint( freelist_cnt )   ) );
+  pack->acct_uses_read  = acct_uses_join( FD_SCRATCH_ALLOC_APPEND( l, acct_uses_align(), acct_uses_footprint( lg_uses_tbl_sz )) );
+  pack->acct_uses_write = acct_uses_join( FD_SCRATCH_ALLOC_APPEND( l, acct_uses_align(), acct_uses_footprint( lg_uses_tbl_sz )) );
+
+  return pack;
+}
+
+ulong fd_pack_bank_cnt( fd_pack_t * pack ) { return pack->bank_cnt; }
+
+ulong fd_pack_avail_txn_cnt( fd_pack_t * pack ) { return outq_cnt( pack->outq ); }
+
+
+/* Helper function that determines whether a transaction is considered a
```

Comment:
The Solana Labs client just uses port received on to determine if something is a vote. Does that make more sense? Also why is this commented out?